react-context/flow.txt

```diff
@@ -0,0 +1,131 @@
+The query then assigns the userId and hashedpassword properties of the user
+object to their own variables.
+Then argon2 verifies the hashedPassword with the inputted password.
+If they don’t match a 401 error ‘invalid login’ gets returned.
+If they do match it will create a unique token and assign it to a token property,
+and add user to the user property.
+Then the Post method sends the result back to the handleSubmit method.
+The hash route is changed from ‘sign-in’ to ‘’(empty string); this triggers
+componentDidMount which gets the token from local storage, and sets the user
+to the user that matches the token as well as setting isAuthorizing to false.
+Now render gets called again and this time the path is ‘’ which returns the
+Home module and changes the Navbar module.
+In Navbar if the user does not equal null the button in the top right is changed
+to a Sign - Out button.
+The Home module calls a render method that checks if there is not a user logged
+in redirect to the ‘sign-in’ page.
+Since we are logged in the page renders a div that contains a rick roll gif.
+
+3. ** USER CAN STAY SIGNED IN (EVEN IF THEY REFRESH THE PAGE) **
+
+Assuming the user is currently signed in and on the homepage.
+When the page is reloaded it re-renders the page which then triggers
+componentDidMount which looks for a hashchange. Since there was no hashchange
+it looks to get a token out of local storage. Since when we logged in the token
+was set in local storage the browser gets the token and matches it with the user.
+Then componentDidMount will change state to the user that the token matches and
+change the isAuthorizing property to false.
+The state change triggers a re-render and the browser sees there is an empty
+string after the hash in the url which makes renderPage() trigger the Home module.
+The render() method re-renders the home page allowing the user to stay on the
+home page when refreshing the page.
+
+4. ** USER CAN SIGN OUT **
+
+Assuming the user is currently signed in and on the homepage.
+Since there is a user signed in Navbar creates a Sign Out button on the top right.
+Once we’re ready to sign out we click the sign out button.
+Once the button is clicked it triggers the handleSignOut method in app.jsx
+handleSignOut will remove the user token that is held in local storage and
+call setState to update the user to null.
+In home.jsx if there is no user logged in it calls redirect.jsx and changes
+the hash url to ‘sign-in’.
+Back in app.jsx the handleSignOut calls for a re-render of the page, since the
+path is now equal to ‘sign-in’ renderPage() returns the Auth module, which returns
+the sign in page. render() then renders the sign in page.
+After the render() method componentDidMount() gets called again and updates the
+route property to the hash route ‘sign-in’. Since the token gets removed in
+handleSignOut the user is set null, and isAuthorizing is set to false.
+Now the page is updated and waiting for another change.
```

Comment:
Once App is rendered for the first time, `componentDidMount` is only called _once_. A component's lifecycle can be divided into three "phases": mounting, updating, and unmounting.

`root.render(<App />)` "mounts" the App component and it's constructor -> render -> componentDidMount methods are called, in that order, but `componentDidMount()` only runs _one_ time. This **completes** the mounting phase of the component. After that, it can only go through runs of its "updating" phase (maybe over and over). Often these update cycles are triggered by `setState`. In the case of App, it's never unmounted, so it stays in its updating phase until the user leaves.

However, the _pages_ will all go through all phases, mounting, updating, and unmounting. When the haschange event occurs, it triggers App's update cycle, but if a new page is rendered, the previous page goes through "unmounting" and the new page starts over with "mounting".

https://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/